hikari/api/chunker.py

```diff
@@ -28,8 +28,131 @@
 import abc
 import typing
 
+from hikari import undefined
+
 if typing.TYPE_CHECKING:
+    import datetime
+
     from hikari import guilds
+    from hikari import snowflakes
+    from hikari import users as users_
+    from hikari.api import shard as gateway_shard
+    from hikari.events import shard_events
+    from hikari.utilities import event_stream
+
+
+class RequestInformation(typing.Protocol):
+    """Information about a member request that's being tracked.
+
+    This protocol defines the fields that should be exported by a `GuildChunker`
+    implementation when getting the tracked information about a request.
+    """
+
+    __slots__: typing.Sequence[str] = ()
+
+    @property
+    def average_chunk_size(self) -> typing.Optional[int]:
+        """Average amount of members that are being received per chunk.
+
+        Returns
+        -------
+        builtins.int or builtins.None
+            The `builtins.int` average size of each chunk for this request
+            or `builtins.None` if we haven't received a response to pull
+            information from yet.
+        """
+
+    @property
+    def chunk_count(self) -> typing.Optional[int]:
+        """Amount of chunks that are expected for this request.
+
+        Returns
+        -------
+        builtins.int or builtins.None
+            The `builtins.int` count of how many chunk events should be received
+            for this request or `builtins.None` if we haven't received a
+            response to pull information from yet.
+        """
+
+    @property
+    def guild_id(self) -> snowflakes.Snowflake:
+        """Snowflake ID of the guild this chunk request is for.
+
+        Returns
+        -------
+        hikari.snowflake.Snowflake
+            The ID of the guild this request is for.
+        """
+
+    @property
+    def is_complete(self) -> bool:
+        """Whether this chunk request is finished or not.
+
+        A chunk request may be considered finished after all chunks have been
+        received or after it's timed out.
+
+        Returns
+        -------
+        builtins.bool
+            Whether this chunk request is considered to be finished or not yet.
+        """
+
+    @property
+    def last_received(self) -> typing.Optional[datetime.datetime]:
+        """Datetime of when we last received a chunk for this event.
+
+        Returns
+        -------
+        datetime.datetime or builtins.None
```

Comment:
We are using typing.Optional over `or` now, as it doesn't break pdoc in some edge cases as much.